pyartifactory/objects/artifact.py

```diff
@@ -0,0 +1,326 @@
+
+from pyartifactory.exception import ArtifactNotFoundError, ArtifactoryError, BadPropertiesError, PropertyNotFoundError
+from pyartifactory.models.artifact import (
+    ArtifactFileInfoResponse,
+    ArtifactFolderInfoResponse,
+    ArtifactInfoResponse,
+    ArtifactListResponse,
+    ArtifactPropertiesResponse,
+    ArtifactStatsResponse,
+)
+from pyartifactory.objects.object import ArtifactoryObject
+
+logger = logging.getLogger("pyartifactory")
+
+
+class ArtifactoryArtifact(ArtifactoryObject):
+    """Models an artifactory artifact."""
+
+    def _walk(self, artifact_path: str, topdown: bool = True) -> Iterator[ArtifactInfoResponse]:
+        """Iterate over artifact (file or directory) recursively.
+
+        :param artifact_path: Path to file or folder in Artifactory
+        :param topdown: True for a top-down (directory first) traversal
+        """
+        info = self.info(artifact_path)
+        if not isinstance(info, ArtifactFolderInfoResponse):
+            yield info
+        else:
+            if topdown:
+                yield info
+            for subdir in (child for child in info.children if child.folder is True):
+                yield from self._walk(artifact_path + subdir.uri, topdown=topdown)
+            for file in (child for child in info.children if child.folder is not True):
+                yield from self._walk(artifact_path + file.uri, topdown=topdown)
+            if not topdown:
+                yield info
+
+    def info(self, artifact_path: Union[Path, str]) -> ArtifactInfoResponse:
+        """
+        Retrieve information about a file or a folder
+
+        See https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-FolderInfo
+        and https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-FileInfo
+
+        :param artifact_path: Path to file or folder in Artifactory
+        """
+
+        if isinstance(artifact_path, str):
+            artifact_path = Path(artifact_path)
+
+        try:
+            response = self._get(f"api/storage/{artifact_path.as_posix()}")
+            model = TypeAdapter(ArtifactInfoResponse).validate_python(response.json())
+            return typing.cast(ArtifactInfoResponse, model)
+        except requests.exceptions.HTTPError as error:
+            if error.response.status_code == 404:
+                logger.error("Artifact %s does not exist", artifact_path)
+                raise ArtifactNotFoundError(f"Artifact {artifact_path} does not exist")
+            raise ArtifactoryError from error
+
+    def deploy(self, local_file_location: Union[Path, str], artifact_path: Union[Path, str]) -> ArtifactInfoResponse:
+        """
+        Deploy a file or directory.
+        :param artifact_path: Path to artifactory in Artifactory
+        :param local_file_location: Location of the file or folder to deploy
+        """
+        local_file = Path(local_file_location)
+        artifact_folder = Path(artifact_path)
```

Comment:
Actually, since we now accept type `Union[Path, str]`, we will need something like:

if isinstance(local_file_location , str):
    local_file = Path(local_file_location )

if isinstance(artifact_path , str):
    artifact_folder = Path(artifact_path )

Reply:
Not really. pathlib supports this case out of the box

![image](https://github.com/anancarv/python-artifactory/assets/16363825/e413a2a3-19b9-4fc8-8aa7-fd8e29ec6fc2)